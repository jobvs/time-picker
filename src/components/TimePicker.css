.widget-time-picker {
    display: flex;
    outline: none;
}

.widget-time-picker-input {
    flex: 1 0 auto;
}

/* Input height is different on mobile devices */
@media only screen and (-webkit-min-device-pixel-ratio: 0) and (max-device-width: 1024px) {
    .widget-time-picker-input input {
        min-height: 38px;
        line-height: 20px;
    }
}

.widget-time-picker-button {
    flex: 0 1 auto;
    margin-left: 5px;
}

.widget-time-picker .rdtPicker {
    width: auto;
    font-size: 12px;
    border: 1px solid #d1d6e4;
    border-radius: 4px;
    box-shadow: 0 2px 10px 0 rgba(0, 0, 0, 0.06);
    padding-left: 16px;
    padding-right: 16px;
    margin-top: 0;
}

.widget-time-picker .rdtBtn {
    color: #0595db;
    font-size: 11px;
}
