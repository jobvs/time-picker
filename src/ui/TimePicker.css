.widget-timepicker {
  display: flex;
}

.widget-timepicker-input {
  flex: 1 0 auto;
}

.widget-timepicker-button {
  flex: 0 1 auto;
  margin-left: 5px;
}