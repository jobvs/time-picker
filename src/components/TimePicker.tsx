import moment from "moment";
import * as React from "react";
import Datetime from "react-datetime";
import { hot } from "react-hot-loader";

import "react-datetime/css/react-datetime.css";

import "../ui/TimePicker.css";
import { Label } from "./Label";

export interface Props {
    name: string;
    class?: string;
    style?: React.CSSProperties;
    tabIndex?: number;

    showLabel: boolean;
    labelCaption: string;
    labelOrientation: "horizontal" | "vertical";
    labelWidth: number;

    timeFormat: "minutes" | "seconds" | "milliseconds";
    timeNotation: "h12" | "h24";
    inputValue?: PluginWidget.EditableValue<Date | undefined>;
    editable: "default" | "never";
    placeholder: PluginWidget.DynamicValue<string>;

    hoursStep?: number;
    minutesStep?: number;
    secondsStep?: number;

    onChange?: PluginWidget.ActionValue;
    onEnter?: PluginWidget.ActionValue;
    onLeave?: PluginWidget.ActionValue;
}

interface State {
    invalidInput: boolean;
}

export class TimePicker extends React.Component<Props, State> {
    state: State = {
        invalidInput: false
    };

    constructor(props: Props) {
        super(props);
        this.handleChange = this.handleChange.bind(this);
        this.handleFocus = this.handleFocus.bind(this);
        this.handleBlur = this.handleBlur.bind(this);
    }

    handleFocus() {
        if (this.props.onEnter && this.props.onEnter.canExecute) {
            this.props.onEnter.execute();
        }
    }

    handleBlur() {
        if (this.state.invalidInput && this.props.inputValue) {
            this.props.inputValue.setValidation("Invalid time");
            this.setState({ invalidInput: true });
        }

        if (this.props.onLeave && this.props.onLeave.canExecute) {
            this.props.onLeave.execute();
        }
    }

    handleChange(input: string | moment.Moment) {
        if (this.props.inputValue === undefined) {
            return;
        }

        if (typeof input === "string" && input.length > 0) {
            this.setState({ invalidInput: true });
            return;
        }

        const value = typeof input === "string" ? undefined : input.toDate();
        this.props.inputValue.setValidation();
        this.props.inputValue.setValue(value);
        this.setState({ invalidInput: false });

        if (this.props.onChange && this.props.onChange.canExecute) {
            this.props.onChange.execute();
        }
    }

    render() {
        return (
            <div className={this.props.class} style={this.props.style}>
                {this.props.showLabel ? this.renderInputWithLabel() : this.renderInput()}
            </div>
        );
    }

    renderInputWithLabel() {
        const hasError =
            this.props.inputValue && this.props.inputValue.validation
                ? this.props.inputValue.validation.length > 0
                : false;

        return (
            <Label
                label={this.props.labelCaption}
                orientation={this.props.labelOrientation}
                width={this.props.labelWidth}
                hasError={hasError}
            >
                {this.renderInput()}
            </Label>
        );
    }

    renderInput() {
        const placeholder = !this.props.inputValue ? "No attribute selected" : this.props.placeholder.value;
        const disabled = this.props.editable === "never" || !this.props.inputValue || this.props.inputValue.readOnly;
        const validation = (this.props.inputValue && this.props.inputValue.validation) || [];
        const value = this.props.inputValue && this.props.inputValue.value;

        return (
            <>
                <div className="widget-timepicker">
                    <Datetime
                        value={value}
                        onChange={this.handleChange}
                        onFocus={this.handleFocus}
                        onBlur={this.handleBlur}
                        dateFormat={false}
                        timeFormat={timeFormat(this.props.timeNotation, this.props.timeFormat)}
                        timeConstraints={timeConstraints(
                            this.props.hoursStep,
                            this.props.minutesStep,
                            this.props.secondsStep
                        )}
                        inputProps={{
                            className: "form-control",
                            name: this.props.name,
                            tabIndex: this.props.tabIndex,
                            placeholder,
                            disabled
                        }}
                    />
                    <span className="glyphicon glyphicon-time form-control-feedback" aria-hidden="true" />
                </div>
                {validation.map(message => (
                    <div key={message} className="alert alert-danger mx-validation-message">
                        {message}
                    </div>
                ))}
            </>
        );
    }
}

function timeConstraints(hoursStep?: number, minutesStep?: number, secondsStep?: number) {
    return {
        hours: { step: hoursStep || 1, min: 0, max: 23 },
        minutes: { step: minutesStep || 1, min: 0, max: 59 },
        seconds: { step: secondsStep || 1, min: 0, max: 59 }
    };
}

function timeFormat(notation: "h12" | "h24", detail: "minutes" | "seconds" | "milliseconds") {
    const hourFormat = notation === "h12" ? "hh" : "HH";
    const seconds = detail === "seconds" ? ":ss" : "";
    const milliseconds = detail === "milliseconds" ? ":ss.SSS" : "";
    const suffix = notation === "h12" ? " A" : "";

    return `${hourFormat}:mm${seconds}${milliseconds}${suffix}`;
}

// tslint:disable-next-line:no-default-export
export default hot(module)(TimePicker);
