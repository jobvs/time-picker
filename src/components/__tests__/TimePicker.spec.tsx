import { shallow } from "enzyme";
import * as React from "react";

import TimePicker, { TimePickerProps } from "../TimePicker";

describe("TimePicker", () => {
    const createTimePicker = (props: TimePickerProps) => shallow(<TimePicker {...props} />);

    it("should render the structure", () => {
        const timepickerProps: TimePickerProps = {
            timepickerType: "badge",
            bootstrapStyle: "default",
            onClickAction: jasmine.createSpy("onClick"),
            value: "0"
        };
        const timepicker = createTimePicker(timepickerProps);

        expect(timepicker).toBeElement(
            <span className={("widget-timepicker badge label-default")}
                onClick={jasmine.any(Function) as any}
                style={timepickerProps.style}>0</span>
        );
    });

    it("should show no value when no value or default value provided", () => {
        const value = "value";
        const timepicker = createTimePicker({ timepickerType: "label", value, defaultValue: "default value" });

        expect(timepicker.text()).toBe(value);
    });

    it("should show default value when no value is provided", () => {
        const defaultValue = "default";
        const timepicker = createTimePicker({ timepickerType: "label", value: undefined, defaultValue });

        expect(timepicker.text()).toBe(defaultValue);
    });

    it("should show no value when no value or default value provided", () => {
        const timepicker = createTimePicker({ timepickerType: "label", value: undefined });

        expect(timepicker.text()).toBe("");
    });

    it("configured as a label should have the class label", () => {
        const timepicker = createTimePicker({ timepickerType: "label" });

        expect(timepicker.hasClass("label")).toBe(true);
    });

    it("configured as a badge should have the class badge", () => {
        const timepicker = createTimePicker({ timepickerType: "badge" });

        expect(timepicker.hasClass("badge")).toBe(true);
    });

    it("with a click action should respond to click events", () => {
        const timepickerProps: TimePickerProps = { onClickAction: jasmine.createSpy("onClick"), timepickerType: "badge" };
        const onClick = timepickerProps.onClickAction = jasmine.createSpy("onClick");
        const timepicker = createTimePicker(timepickerProps);

        timepicker.simulate("click");

        expect(onClick).toHaveBeenCalledTimes(1);
    });

    it("with the Bootstrap style default should have the class label-default", () => {
        const timepicker = createTimePicker({ bootstrapStyle: "default", timepickerType: "badge" });

        expect(timepicker.hasClass("label-default")).toBe(true);
    });

    it("with the Bootstrap style primary should have the class label-primary", () => {
        const timepicker = createTimePicker({ bootstrapStyle: "primary", timepickerType: "badge" });

        expect(timepicker.hasClass("label-primary")).toBe(true);
    });

    it("with the Bootstrap style success should have the class label-success", () => {
        const timepicker = createTimePicker({ bootstrapStyle: "success", timepickerType: "badge" });

        expect(timepicker.hasClass("label-success")).toBe(true);
    });

    it("with the Bootstrap style info should have the class label-info", () => {
        const timepicker = createTimePicker({ bootstrapStyle: "info", timepickerType: "badge" });

        expect(timepicker.hasClass("label-info")).toBe(true);
    });

    it("with the Bootstrap style warning should have the class label-warning", () => {
        const timepicker = createTimePicker({ bootstrapStyle: "warning", timepickerType: "badge" });

        expect(timepicker.hasClass("label-warning")).toBe(true);
    });

    it("with the Bootstrap style danger should have the class label-danger", () => {
        const timepicker = createTimePicker({ bootstrapStyle: "danger", timepickerType: "badge" });

        expect(timepicker.hasClass("label-danger")).toBe(true);
    });
});
