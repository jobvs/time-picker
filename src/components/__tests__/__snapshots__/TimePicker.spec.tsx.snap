// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimePicker label should render a horizontal label 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="form-group clearfix "
  >
    <label
      className="control-label col-sm-3"
    >
      Default caption
    </label>
    <div
      className="col-sm-9"
    >
      <div
        className="widget-timepicker"
      >
        <div
          className="rdt"
        >
          <input
            className="form-control"
            disabled={false}
            name="custom-name"
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Placeholder"
            tabIndex={0}
            type="text"
            value="01:02:03.456 AM"
          />
          <div
            className="rdtPicker"
          >
            <div
              className="rdtTime"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <div
                        className="rdtCounters"
                      >
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            1
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            02
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            03
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            AM
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter rdtMilli"
                        >
                          <input
                            onChange={[Function]}
                            type="text"
                            value="456"
                          />
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-time form-control-feedback"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimePicker label should render a horizontal label with a specified width 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="form-group clearfix "
  >
    <label
      className="control-label col-sm-6"
    >
      Default caption
    </label>
    <div
      className="col-sm-6"
    >
      <div
        className="widget-timepicker"
      >
        <div
          className="rdt"
        >
          <input
            className="form-control"
            disabled={false}
            name="custom-name"
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Placeholder"
            tabIndex={0}
            type="text"
            value="01:02:03.456 AM"
          />
          <div
            className="rdtPicker"
          >
            <div
              className="rdtTime"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <div
                        className="rdtCounters"
                      >
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            1
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            02
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            03
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            AM
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter rdtMilli"
                        >
                          <input
                            onChange={[Function]}
                            type="text"
                            value="456"
                          />
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-time form-control-feedback"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimePicker label should render a vertical label 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="form-group clearfix "
  >
    <label
      className="control-label "
    >
      Default caption
    </label>
    <div
      className=""
    >
      <div
        className="widget-timepicker"
      >
        <div
          className="rdt"
        >
          <input
            className="form-control"
            disabled={false}
            name="custom-name"
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Placeholder"
            tabIndex={0}
            type="text"
            value="01:02:03.456 AM"
          />
          <div
            className="rdtPicker"
          >
            <div
              className="rdtTime"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <div
                        className="rdtCounters"
                      >
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            1
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            02
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            03
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            AM
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter rdtMilli"
                        >
                          <input
                            onChange={[Function]}
                            type="text"
                            value="456"
                          />
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-time form-control-feedback"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimePicker label should render with a different caption 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="form-group clearfix "
  >
    <label
      className="control-label "
    >
      Different
    </label>
    <div
      className=""
    >
      <div
        className="widget-timepicker"
      >
        <div
          className="rdt"
        >
          <input
            className="form-control"
            disabled={false}
            name="custom-name"
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            placeholder="Placeholder"
            tabIndex={0}
            type="text"
            value="01:02:03.456 AM"
          />
          <div
            className="rdtPicker"
          >
            <div
              className="rdtTime"
            >
              <table>
                <tbody>
                  <tr>
                    <td>
                      <div
                        className="rdtCounters"
                      >
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            1
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            02
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            03
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounter"
                        >
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▲
                          </span>
                          <div
                            className="rdtCount"
                          >
                            AM
                          </div>
                          <span
                            className="rdtBtn"
                            onContextMenu={[Function]}
                            onMouseDown={[Function]}
                          >
                            ▼
                          </span>
                        </div>
                        <div
                          className="rdtCounterSeparator"
                        >
                          :
                        </div>
                        <div
                          className="rdtCounter rdtMilli"
                        >
                          <input
                            onChange={[Function]}
                            type="text"
                            value="456"
                          />
                        </div>
                      </div>
                    </td>
                  </tr>
                </tbody>
              </table>
            </div>
          </div>
        </div>
        <span
          aria-hidden="true"
          className="glyphicon glyphicon-time form-control-feedback"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`TimePicker should render 24-hour notation 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="widget-timepicker"
  >
    <div
      className="rdt"
    >
      <input
        className="form-control"
        disabled={false}
        name="custom-name"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Placeholder"
        tabIndex={0}
        type="text"
        value="01:02:03.456"
      />
      <div
        className="rdtPicker"
      >
        <div
          className="rdtTime"
        >
          <table>
            <tbody>
              <tr>
                <td>
                  <div
                    className="rdtCounters"
                  >
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        1
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        02
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        03
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter rdtMilli"
                    >
                      <input
                        onChange={[Function]}
                        type="text"
                        value="456"
                      />
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <span
      aria-hidden="true"
      className="glyphicon glyphicon-time form-control-feedback"
    />
  </div>
</div>
`;

exports[`TimePicker should render the structure 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="widget-timepicker"
  >
    <div
      className="rdt"
    >
      <input
        className="form-control"
        disabled={false}
        name="custom-name"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="Placeholder"
        tabIndex={0}
        type="text"
        value="01:02:03.456 AM"
      />
      <div
        className="rdtPicker"
      >
        <div
          className="rdtTime"
        >
          <table>
            <tbody>
              <tr>
                <td>
                  <div
                    className="rdtCounters"
                  >
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        1
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        02
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        03
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        AM
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter rdtMilli"
                    >
                      <input
                        onChange={[Function]}
                        type="text"
                        value="456"
                      />
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <span
      aria-hidden="true"
      className="glyphicon glyphicon-time form-control-feedback"
    />
  </div>
</div>
`;

exports[`TimePicker should render without a data attribute 1`] = `
<div
  className="custom-class"
  style={
    Object {
      "backgroundColor": "transparent",
    }
  }
>
  <div
    className="widget-timepicker"
  >
    <div
      className="rdt"
    >
      <input
        className="form-control"
        disabled={true}
        name="custom-name"
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="No attribute selected"
        tabIndex={0}
        type="text"
        value=""
      />
      <div
        className="rdtPicker"
      >
        <div
          className="rdtTime"
        >
          <table>
            <tbody>
              <tr>
                <td>
                  <div
                    className="rdtCounters"
                  >
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        12
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        00
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        00
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounter"
                    >
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▲
                      </span>
                      <div
                        className="rdtCount"
                      >
                        AM
                      </div>
                      <span
                        className="rdtBtn"
                        onContextMenu={[Function]}
                        onMouseDown={[Function]}
                      >
                        ▼
                      </span>
                    </div>
                    <div
                      className="rdtCounterSeparator"
                    >
                      :
                    </div>
                    <div
                      className="rdtCounter rdtMilli"
                    >
                      <input
                        onChange={[Function]}
                        type="text"
                        value="000"
                      />
                    </div>
                  </div>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
    <span
      aria-hidden="true"
      className="glyphicon glyphicon-time form-control-feedback"
    />
  </div>
</div>
`;
